processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java

```diff
@@ -51,25 +60,44 @@ public SelectorDimFilter(
   {
     byte[] dimensionBytes = StringUtils.toUtf8(dimension);
     byte[] valueBytes = (value == null) ? new byte[]{} : StringUtils.toUtf8(value);
+    byte[] extractionFnBytes = extractionFn == null ? new byte[0] : extractionFn.getCacheKey();
 
-    return ByteBuffer.allocate(2 + dimensionBytes.length + valueBytes.length)
+    return ByteBuffer.allocate(3 + dimensionBytes.length + valueBytes.length + extractionFnBytes.length)
                      .put(DimFilterCacheHelper.SELECTOR_CACHE_ID)
                      .put(dimensionBytes)
                      .put(DimFilterCacheHelper.STRING_SEPARATOR)
                      .put(valueBytes)
+                     .put(DimFilterCacheHelper.STRING_SEPARATOR)
+                     .put(extractionFnBytes)
                      .array();
   }
 
   @Override
   public DimFilter optimize()
   {
+    if (this.getExtractionFn() instanceof LookupExtractionFn
+        && ((LookupExtractionFn) this.getExtractionFn()).isOptimize()) {
+      LookupExtractor lookup = ((LookupExtractionFn) this.getExtractionFn()).getLookup();
+      final List<String> keys = lookup.unapply(this.getValue());
+      final String dimensionName = this.getDimension();
+      if (!keys.isEmpty()) {
```

Comment:
Actually there might be bigger problems?

This is optimizing `lookup(dim) == val` into `dim IN lookup.unapply(val)` but that doesn't properly take into account the missing value handling. e.g. if retainMissingValues is true, and `lookup("val")` were null, then a literal `"val"` should match `lookup(dim) == val` but it won't match `dim IN lookup.unapply(val)`.